### PR TITLE
Fix sync failure state exception handling

### LINKED ISSUE #1654: Download fails to complete when exception is encountered
**Describe the bug**
When an exception is occurs while downloading a resource, the `com.google.android.fhir.sync.download.DownloaderImpl` attempts to emit a `com.google.android.fhir.sync.DownloadState.Failure` failure in the `catch` block which results to a FATAL_EXCEPTION.  

**Component**
Core library

**To Reproduce**
1. Attempt downloading a resource that has an exception. (Example in our use case was an `Encounter` resource missing a `coding.code`)

**Expected behavior**
In case of a Resource sync failure `com.google.android.fhir.sync.DownloadState.Failure` should be emitted with the parameter `ResourceSyncException` wrapping the sync exception without violating the flow exception transparency.

**Error logs**
```java
FATAL EXCEPTION: main
                                                                                                    Process: org.smartregister.fhircore.ecbis, PID: 10487
                                                                                                    java.lang.IllegalStateException: Flow exception transparency is violated:
                                                                                                        Previous 'emit' call has thrown exception java.lang.NullPointerException: coding.code must not be null, but then emission attempt of value 'Failure(syncError=ResourceSyncException(resourceType=Encounter, exception=java.lang.NullPointerException: coding.code must not be null))' has been detected.
                                                                                                        Emissions from 'catch' blocks are prohibited in order to avoid unspecified behaviour, 'Flow.catch' operator can be used instead.
                                                                                                        For a more detailed explanation, please refer to Flow documentation.
                                                                                                    	at kotlinx.coroutines.flow.internal.SafeCollector.exceptionTransparencyViolated(SafeCollector.kt:140)
                                                                                                    	at kotlinx.coroutines.flow.internal.SafeCollector.checkContext(SafeCollector.kt:104)
                                                                                                    	at kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:83)
                                                                                                    	at kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:66)
                                                                                                    	at com.google.android.fhir.sync.download.DownloaderImpl$download$2.invokeSuspend(DownloaderImpl.kt:56)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                                                                                                    	at kotlinx.coroutines.flow.internal.SafeCollector.invokeSuspend(SafeCollector.kt:48)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                                                                                                    	at kotlinx.coroutines.flow.internal.SafeCollector.invokeSuspend(SafeCollector.kt:48)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)
                                                                                                    	at android.os.Handler.handleCallback(Handler.java:938)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:99)
                                                                                                    	at android.os.Looper.loop(Looper.java:223)
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:7656)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
                                                                                                    	Suppressed: kotlinx.coroutines.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@1662c1a, Dispatchers.Main.immediate]

```
Libraries used:
`org.jetbrains.kotlinx:kotlinx-coroutines-core:1.6.2`
`org.jetbrains.kotlinx:kotlinx-coroutines-android:1.6.2`

### PR DESCRIPTION
**Pull request by ellykits**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1654

**Description**
Clear and concise code change description. 

**Type**
Choose one: Bug fix

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [ ] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [ ] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [ ] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

File: demo/src/main/java/com/google/android/fhir/demo/data/DownloadWorkManagerImpl.kt [hunk 1]
```diff
@@ -30,16 +30,17 @@ import org.hl7.fhir.r4.model.ResourceType
 class DownloadWorkManagerImpl : DownloadWorkManager {
   private val resourceTypeList = ResourceType.values().map { it.name }
   private val urls = LinkedList(listOf("Patient?address-city=NAIROBI"))
+  override var nextRequestUrl: String? = null
```
Comment (maimoonak) [inline demo/src/main/java/com/google/android/fhir/demo/data/DownloadWorkManagerImpl.kt:33]: can we make it private. ideally this should not be modifiable outside this class

File: engine/src/main/java/com/google/android/fhir/impl/FhirEngineImpl.kt [hunk 2]
```diff
@@ -104,6 +107,10 @@ internal class FhirEngineImpl(private val database: Database, private val contex
           saveResolvedResourcesToDatabase(resolved)
         }
       }
+      .catch { throwable: Throwable ->
+        Timber.e(throwable, "Error saving remote resource to database")
+      }
+      .collect()
```
Comment (maimoonak) [inline engine/src/main/java/com/google/android/fhir/impl/FhirEngineImpl.kt:113]: we are using on each and doing the processing above. do we need collect now?

Comment (ellykits) [reply engine/src/main/java/com/google/android/fhir/impl/FhirEngineImpl.kt:113]: Yes `collect` operator is required to process all emitted values and handle exceptions that might occur in the upstream flow or during processing.

File: demo/src/main/java/com/google/android/fhir/demo/data/DownloadWorkManagerImpl.kt (hunk 1 above)
Comment (ellykits) [reply demo/src/main/java/com/google/android/fhir/demo/data/DownloadWorkManagerImpl.kt:33]: I see. This was exposed to be used in `com.google.android.fhir.sync.download.DownloaderImpl#download` to get the resource type that is currently being synced. The `catch` operator can only access the `flow` instance and not the internal states within the `flow { }` closure.

Comment (ellykits) [reply demo/src/main/java/com/google/android/fhir/demo/data/DownloadWorkManagerImpl.kt:33]: Do we need the resource type in `ResourceSyncException` data class? Most instance creation of that class have `ResourceType.Bundle` provided as the resource type.

Comment (maimoonak) [reply demo/src/main/java/com/google/android/fhir/demo/data/DownloadWorkManagerImpl.kt:33]: I guess it does emit the failed resource type. the test DownloaderImpl.`downloader with patient and observations should continue to download observations if patient download fail` has an assertion for this.

Comment (ellykits) [reply demo/src/main/java/com/google/android/fhir/demo/data/DownloadWorkManagerImpl.kt:33]: Investigating the failing tests.

Comment (ellykits) [reply demo/src/main/java/com/google/android/fhir/demo/data/DownloadWorkManagerImpl.kt:33]: I figured the cause of the filing tests. So apparently the catch operator completes the flow thus the remaining states are not emitted.

Comment (ellykits) [reply demo/src/main/java/com/google/android/fhir/demo/data/DownloadWorkManagerImpl.kt:33]: The test was verifying whether emissions still happens even when there is a failure. Which I think should be the correct behavior. I will revert and remove the emission in the catch section of the try-catch block.

**Comment by ellykits** [comment]
@maimoonak I will have to revert the commits; then create a new PR instead of maintaining the commit history of reverted code.

**Pull request closed without merging**